sw/host/opentitanlib/src/util/present.rs

```diff
@@ -0,0 +1,274 @@
+// Copyright lowRISC contributors.
+// Licensed under the Apache License, Version 2.0, see LICENSE for details.
+// SPDX-License-Identifier: Apache-2.0
+
+use anyhow::{bail, Result};
+use std::convert::TryInto;
+use zerocopy::AsBytes;
+
+/// PRESENT block cipher.
+///
+/// Based on version 1.2 of the following Python implementation
+/// https://github.com/doegox/python-cryptoplus
+pub struct Present {
+    round_keys: Vec<u64>,
+}
+
+impl Present {
+    pub fn try_new_rounds(key: Vec<u8>, rounds: usize) -> Result<Present> {
+        if rounds < 1 || rounds > 254 {
+            bail!("unsupported number of rounds {}", rounds);
+        }
+
+        let round_keys = match key.len() {
+            10 => generate_round_keys_80(key, rounds),
+            16 => generate_round_keys_128(key, rounds),
+            _ => bail!("key length must be 80 or 128 bits"),
+        };
+
+        Ok(Present { round_keys })
+    }
+
+    /// Create a new instance of the PRESENT cipher.
+    ///
+    /// Valid key lengths are 80 and 128 bits. All other key lengths will return an error.
+    pub fn try_new(key: Vec<u8>) -> Result<Present> {
+        Self::try_new_rounds(key, 32)
+    }
+
+    /// Create a new 128-bit PRESENT cipher instance.
+    pub fn new_128(key: &[u8; 16]) -> Present {
+        Self::try_new(key.to_vec()).unwrap()
+    }
+
+    /// Create a new 80-bit PRESENT cipher instance.
+    pub fn new_80(key: &[u8; 10]) -> Present {
+        Self::try_new(key.to_vec()).unwrap()
+    }
+
+    /// Encrypt a 64-bit block.
+    pub fn encrypt_block(&self, block: u64) -> u64 {
+        let mut state = block;
+        state ^= self.round_keys[0];
+        for round_key in &self.round_keys[1..] {
+            state = s_box_layer(state);
+            state = p_box_layer(state);
+            state ^= round_key;
+        }
+        state
+    }
+
+    /// Decrypt a 64-bit block.
+    pub fn decrypt_block(&self, block: u64) -> u64 {
+        let mut state = block;
+        for round_key in self.round_keys[1..].iter().rev() {
+            state ^= round_key;
+            state = p_box_layer_dec(state);
+            state = s_box_layer_dec(state);
+        }
+        state ^ self.round_keys[0]
+    }
+}
+
+const S_BOX: [u8; 16] = [
+    0x0c, 0x05, 0x06, 0x0b, 0x09, 0x00, 0x0a, 0x0d, 0x03, 0x0e, 0x0f, 0x08, 0x04, 0x07, 0x01, 0x02,
+];
+
+const S_BOX_INV: [u8; 16] = [
+    0x05, 0x0e, 0x0f, 0x08, 0x0c, 0x01, 0x02, 0x0d, 0x0b, 0x04, 0x06, 0x03, 0x00, 0x07, 0x09, 0x0a,
+];
+
+const P_BOX: [u8; 64] = [
```

Comment:
Looking at this p-box, i gotta say, I am soooo glad this doesn't have to be constant-time.